### ISSUE TITLE
package search with pypi fails when package name doesn't match install name

### ISSUE THREAD

**Issue by d0c-s4vage** [d0c-s4vage/pipless#5]
E.g. `pip install Flask` vs `import flask`

**Comment by d0c-s4vage** [comment]
After a lot of research, there just isn't a reliable way to do this besides exact matches (which even then isn't 100% reliable).

PyPi hosts _distributions_ that can be installed through `pip`. A distribution can provide zero or more packages, not necessarily even with the same package namespace. E.g. `pip install Flask`.

You could even have an evil python developer who creates two distributions: `malicious_a` and `malicious_b`. The `malicious_a` _distribution_ could provide a python package `malicious_b`, and the `malicious_b` _distribution_ could provide a python package `malicious_a`. Hopefully neither of those packages would ever become popular.

I should do that just for fun though :^)

**Comment by d0c-s4vage** [comment]
wonder how long it would take to compile a mapping of the top 10000 python packages to their install names

**Comment by d0c-s4vage** [comment]
I've decided to compile a mapping of the top 10,000 or 100,000 python packages to included with pipless. It's the only way to be sure when auto-installing packages.

**Comment by zachriggle** [comment]
PyPi exposes a "provides" field which might work in this scenario, but it requires package maintainers to be diligent and upload a PKG-INFO file.

**Comment by zachriggle** [comment]
You might also run into issues with conflicting names :(

**Comment by d0c-s4vage** [comment]
yeah, the "provides" field isn't consistent either. The general consensus, even among the guys who do pypi development, is that you'd have to install it to be 100% sure. If a wheel was available you wouldn't have to fully install though.

**Comment by zachriggle** [comment]
I feel like that's a major failure of the entire system. "You have to give
me code execution to be sure."

**Comment by d0c-s4vage** [comment]
Perhaps use pipreqs to resolve version numbers?
